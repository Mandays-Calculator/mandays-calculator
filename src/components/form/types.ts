import type { ReactElement, ReactNode } from "react";

export type BaseInputProps<Type = object> = Type & {
  name?: string;
  label?: string | ReactNode;
  error?: boolean;
  helperText?: string | ReactNode;
  isFastField?: boolean;
  readOnly?: boolean;
};

export type WithInputControllerType =
  | "text"
  | "date-picker"
  | "dropzone"
  | "pin-code"
  | "autocomplete"
  | "checkbox-group"
  | "date-time-picker"
  | "number-input";

export type BaseInputPropsWithAdornment<Type = object> = Type & {
  name: string;
  label?: string;
  endAdornment?: ReactElement;
  readOnly?: boolean;
  placeholder?: string;
};

export interface WithInputControllerValueProp {
  value?: string;
  selected?: string;
  startDate?: string;
  endDate?: string;
}

type FormValues = {
  [key: string]: any;
};

export type FormErrors = {
  [key in keyof FormValues]?: string;
};
