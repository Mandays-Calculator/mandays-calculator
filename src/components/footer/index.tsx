export { default as Footer } from "./Footer";
export { StyledFooterContainer, StyledFooterText } from "./styles";
