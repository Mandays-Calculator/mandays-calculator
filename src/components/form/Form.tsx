import type { PropsWithChildren, ReactElement } from "react";
import type { FormikInstance } from "formik";
import { FormikProvider } from "formik";
interface FormProps<FormValues extends object> extends PropsWithChildren {
  instance: FormikInstance<FormValues>;
}

export const Form = <FormValues extends object>(props: FormProps<FormValues>): ReactElement => {
  const { instance, children, ...rest } = props;
  return (
    <FormikProvider value={instance}>
      <form
        {...rest}
        onSubmit={instance.handleSubmit}
      >
        {children}
      </form>
    </FormikProvider>
  );
};

export default Form;
