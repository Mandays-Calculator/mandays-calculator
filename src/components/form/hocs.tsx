import type { ComponentType, ReactElement } from "react";
import type { FormikValues } from "formik";
import type {
  BaseInputProps,
  WithInputControllerType,
  WithInputControllerValueProp,
} from "./types";
import type { ReactDatePickerFunctionParams } from "../global-types/app";

import { useMemo } from "react";
import { useFormikContext, getIn, Field, FastField } from "formik";
import { isArray } from "lodash";

type WrappedInputProps<Type> = Type &
  BaseInputProps & {
    name: string;
  };
type WithInputController<Type extends object> = (props: WrappedInputProps<Type>) => ReactElement;

export const withInputController = <Type extends FormikValues>(
  WrappedInput: ComponentType<Type>,
  inputType: WithInputControllerType = "text"
): WithInputController<Type> => {
  return (props: WrappedInputProps<Type>): ReactElement => {
    const { name, isFastField, ...rest } = props;
    const form = useFormikContext<Type>();

    const fieldValue = form.values[name] || getIn(form.values, name);
    const fieldError = form.errors[name] || getIn(form.errors, name);
    const fieldTouched = form.touched[name] || getIn(form.touched, name);
    const hasError = Boolean(fieldError) && Boolean(fieldTouched);

    const handleInputNumberChange = (_: any, value: number): void => {
      form.setFieldValue(name, value);
    };

    const handleDateChange = (date: ReactDatePickerFunctionParams): void => {
      form.setFieldValue(name, date);
    };

    const handleDropzoneChange = (files: File[]): void => {
      form.setFieldValue(name, files);
    };

    const handlePinCodeChange = (pinCode: string): void => {
      form.setFieldValue(name, pinCode);
    };

    const handleCheckBoxChanges = (checkBoxName: string, checked: boolean): void => {
      if (checked) {
        form.setFieldValue(name, [checkBoxName, ...(form.values[name] ?? [])]);
      } else {
        const checkBoxData: string[] = [...form.values[name]];
        const index = checkBoxData.indexOf(checkBoxName);

        if (index !== -1) {
          checkBoxData.splice(index, 1);
          form.setFieldValue(name, checkBoxData);
        }
      }
    };

    const handleAutoComplete = (data: Type): void => {
      form.setFieldValue(name, data);
    };

    const getInputChange = ():
      | typeof handleDateChange
      | typeof handleDropzoneChange
      | typeof handlePinCodeChange
      | typeof handleCheckBoxChanges
      | typeof handleAutoComplete
      | typeof handleInputNumberChange
      | typeof form.handleChange => {
      switch (inputType) {
        case "date-picker":
          return handleDateChange;
        case "dropzone":
          return handleDropzoneChange;
        case "pin-code":
          return handlePinCodeChange;
        case "checkbox-group":
          return handleCheckBoxChanges;
        case "autocomplete":
          return handleAutoComplete;
        case "number-input":
          return handleInputNumberChange;
        default:
          return form.handleChange(name);
      }
    };

    const getInputValueProp = (): WithInputControllerValueProp => {
      if (inputType === "date-picker") {
        if (isArray(fieldValue)) {
          return {
            startDate: fieldValue[0],
            endDate: fieldValue[1],
          };
        }
        return {
          selected: fieldValue,
        };
      }

      return {
        value: fieldValue,
      };
    };

    const MemoriedField = useMemo(() => (isFastField ? FastField : Field), [isFastField]);
    return (
      <MemoriedField
        component={WrappedInput}
        name={name}
        error={hasError}
        {...getInputValueProp()}
        onChange={getInputChange()}
        {...(rest as Type)}
      />
    );
  };
};
