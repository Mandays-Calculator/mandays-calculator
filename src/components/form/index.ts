export * from "./checkbox";
export * from "./checkbox-group";
export * from "./text-field";
export * from "./date-picker";
