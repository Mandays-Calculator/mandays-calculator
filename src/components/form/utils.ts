import type { FormErrors } from "./types";
import { defaultTo, get } from "lodash";

export const getOption = <Type extends object>(
  path: keyof Type | "value" | "label",
): ((option: Type) => string) => {
  return (option: Type) => {
    return defaultTo(get(option, path), "");
  };
};

export const getFieldError = (
  errors: FormErrors | undefined,
  name: string,
): string | undefined => {
  if (errors) {
    return typeof get(errors, name) === "string" ? get(errors, name) : "";
  }
};
