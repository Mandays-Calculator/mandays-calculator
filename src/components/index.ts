export * from "./form/text-field";
export * from "./form/Form";
export * from "./title";
