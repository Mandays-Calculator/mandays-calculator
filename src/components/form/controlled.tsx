import { CheckBox, CheckboxGroup, DatePicker, SearchSelect, TextField, NumberInput } from ".";
import DateTimePicker from "./date-time-picker/DateTimePicker";
import { withInputController } from "./hocs";
import Select from "./select/Select";
import TextArea from "./text-area/TextArea";

export const ControlledTextField = withInputController(TextField);
export const ControlledCheckBox = withInputController(CheckBox);
export const ControlledTextArea = withInputController(TextArea);
export const ControlledCheckBoxGroup = withInputController(
  CheckboxGroup,
  "checkbox-group"
) as typeof CheckboxGroup;
export const ControlledDatePicker = withInputController(
  DatePicker,
  "date-picker"
);
export const ControlledSelect = withInputController(Select);
export const ControlledSearchSelect = withInputController(
  SearchSelect,
  "autocomplete"
) as typeof SearchSelect;
export const ControlledDateTimePicker = withInputController(
  DateTimePicker,
  "date-time-picker"
);
export const ControlledNumberInput = withInputController(NumberInput, "number-input");
