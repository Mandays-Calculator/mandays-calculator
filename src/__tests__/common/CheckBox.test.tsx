import { render } from "@testing-library/react";
import { CheckBox } from "~/components/form";

describe("CheckBox Component", () => {
  it("renders checkbox without crashing", () => {
    const { getByTestId, asFragment } = render(<CheckBox name="test-checkbox" />);
    const checkbox = getByTestId("checkbox-component");
    expect(checkbox).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders helper text", () => {
    const { getByTestId, asFragment } = render(
      <CheckBox
        name="test-checkbox"
        helperText="helper text"
      />
    );
    const checkbox = getByTestId("checkbox-component");
    expect(checkbox).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
