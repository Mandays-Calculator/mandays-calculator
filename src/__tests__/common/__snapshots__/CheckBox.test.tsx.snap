// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox Component renders checkbox without crashing 1`] = `
<DocumentFragment>
  <fieldset
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      data-testid="checkbox-component"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="test-checkbox"
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          name="test-checkbox"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
    </label>
  </fieldset>
</DocumentFragment>
`;

exports[`CheckBox Component renders helper text 1`] = `
<DocumentFragment>
  <fieldset
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      data-testid="checkbox-component"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="test-checkbox"
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          name="test-checkbox"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
    </label>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
    >
       helper text 
    </p>
  </fieldset>
</DocumentFragment>
`;
